### ISSUE TITLE
Relation between `named_parameters` given to `DistributedOptimizer` and the weight parameters given to the wrapped optimizer

### ISSUE THREAD

**Issue by bankotsu** [horovod/horovod#2181]
**Environment:**
1. Framework: PyTorch
2. Framework version: 1.6.0
3. Horovod version: 0.19.5
4. MPI version:
5. CUDA version:
6. NCCL version:
7. Python version: 3.6
8. OS and version:
9. GCC version:

**Checklist:**
1. Did you search issues to find if somebody asked this question before? yes
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)?
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)?
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)? yes

**Your question:**
When initializing a `DistributedOptimizer` we need to supply a `named_parameters` argument and a `torch.optim` `optimizer` argument. My question is: what should be the relation between `named_parameters` and the `params` argument one initiliazes the `optimizer` with?
for example, in the attached screenshot (taken from https://github.com/horovod/horovod/blob/master/examples/pytorch_mnist.py):
![image](https://user-images.githubusercontent.com/13404522/90192222-2d84d100-ddcb-11ea-90b9-191c5d321546.png)
should `model.parameters()` (marked in red) and `model.named_parameters()` (marked in blue) correspond to the same model weight parameters? or maybe the `named_parameters` should always include all the parameters of the model regardless of the `params` argument given to the `optimizer`?
lets say i supply my optimizer with customized param groups:
```python
adam = torch.optim.Adam([
        {'params': params_1},
        {'params': params_2, 'weight_decay': decay},
        {'params': params_3, 'lr': lr_1},
], lr=lr_2)
```
what should be the `named_parameters` value i supply to the corresponding `DistributedOptimizer` that wraps the optimizer in this case?
should the parameters it contains correspond to each of the weight parameters in `params_1`, `params_2` and `params_3`? or should it simply just be `model.named_parameters()`?

Thanks in advance!

**Comment by tgaddair** [comment]
Hey @bankotsu, sorry for the late response.  To answer your question: the `named_parameters` given to `DistributedOptimizer` should correspond to the same `parameters` given to the original optimizer.  The only reason for the distinction is that when performing the `allreduce` we need to assign every parameter a unique name so different processes know which tensors to average together.

In fact, the `named_parameters` param to `DistributedOptimizer` is optional, and when it's not specified we will simply assign a very generic name to each param [here](https://github.com/horovod/horovod/blob/master/horovod/torch/optimizer.py#L42). This will work in any case where your parameters are in the same order across different workers.

**Comment by bankotsu** [comment]
Thanks @tgaddair for clearing this up!

regarding your second point:
>In fact, the named_parameters param to DistributedOptimizer is optional, and when it's not specified we will simply assign a very generic name to each param here. This will work in any case where your parameters are in the same order across different workers.

i think then that i should report a bug:
if i supply `DistributedOptimizer` with an optimizer initialized with param groups as my example above:
```
optimizer = torch.optim.Adam([
        {'params': params_1},
        {'params': params_2, 'weight_decay': decay},
        {'params': params_3, 'lr': lr_1},
], lr=lr_2)
```
and don't specify the optional `named_parameters` param, it will assign the same names across param groups:
```
named_parameters = [('allreduce.noname.%s' % i, v)
                    for param_group in self.param_groups
                    for i, v in enumerate(param_group['params'])]
```
(same `i` and therefore same name is assigned across param groups)

and this will cause an unexpected error to be raised:
"Parameter names in named_parameters must be unique. Found duplicates: ..."

**Comment by tgaddair** [comment]
Yes, good point, we should  also include the `param_group` index in the name.
